universe/ScriptingContext.h

```diff
@@ -9,6 +9,20 @@
 
 class UniverseObject;
 
+struct BackgroundContext {
+public:
+    BackgroundContext()
+    {}
+
+    BackgroundContext(int bout_, Universe::EmpireObjectVisibilityMap& empire_object_visibility_) :
+        bout(bout_),
+        empire_object_visibility(empire_object_visibility_)
+    {}
+
+    int bout;
+    Universe::EmpireObjectVisibilityMap empire_object_visibility;       ///< indexed by empire id and object id, the visibility level the empire has of each object.  may be increas
```

Comment:
you're storing a copy of the input map here. it should be possible to store a reference or pointer which points to a static empty map if not specified in the constructor.

Reply:
the main problem is that i still stored a copy of the background in each ScriptingContext.
Changed that to a reference and tried to create an empty background only once and reuse that if no special info is given  (already pushed).

Reply:
This still makes a copy of the visibility map whenever you construct a background context. Should store it by reference as well... and probably just get rid of the background context and store directly in the scripting context.

Reply:
> This still makes a copy of the visibility map whenever you construct a background context. Should store it by reference as well... and probably just get rid of the background context and store directly in the scripting context.

Uhm sounds like you want the original implementation? You told me you do not want more things in the scripting context. So where should the bout go?

Reply:
Ideal would perhaps be moving the combat round tracking into combat info or autoresolve info, which also has a visibility map in it, and then pass the info struct to the scripting context constructor and store reference to it while also extracting and using the vis info to set a separate reference in scripting context. But doing that requires having a constructor for scripting context that takes a combat info struct of some sort, which might require putting that struct in common... So that can maybe be worked around by making a base class for combat info that contains the combat round and the vis info only, and putting that in common but not the full combat info. So hopefully then you could initialize a scripting context with a combat info (including combat round) without the extra stuff going into the common library.

But at this point I thought for just this issue, it's probably fine to have the combat round stored by value in the scripting context directly and worry about all the rest uf and when more combat info type stuff is to be made accessible from the context. More important is to avoid needing to copy and initialize whole new empty visibility maps with every scripting context and to avoid copying them unnecessarily when passing one to create a scripting context.

Reply:
I tried with AutoresolveInfo before doing the struct. In order to get anywhere i had to expose tons of structure and had add dummy creation ... so i said that is flawed do something simple

See the last commit in [agrrr3/freeorion/202001_CombatBoutCondition_09_stab_at_AutoresolveInfo](https://github.com/agrrr3/freeorion/commits/202001_CombatBoutCondition_09_stab_at_AutoresolveInfo) ( fecb3f0525d3bf2f10c90ca2cbc8a4ef8245a20f )

Reply:
That's why I mentioned making a base class that is public that AutoresolveInfo derives from, so you wouldn't need to make the full contents of AutoresolveInfo public...

Reply:
Also did not work, I went for the CombatInfo instead.